packages/blocks/src/api/factory.js

```diff
@@ -345,11 +408,18 @@ export function switchToBlockType( blocks, name ) {
 	}
 
 	let transformationResults;
+
 	if ( transformation.isMultiBlock ) {
-		transformationResults = transformation.transform(
-			blocksArray.map( ( currentBlock ) => currentBlock.attributes ),
-			blocksArray.map( ( currentBlock ) => currentBlock.innerBlocks )
-		);
+		if ( has( transformation, 'apply' ) ) {
+			transformationResults = transformation.apply( blocksArray );
+		} else {
+			transformationResults = transformation.transform(
+				blocksArray.map( ( currentBlock ) => currentBlock.attributes ),
+				blocksArray.map( ( currentBlock ) => currentBlock.innerBlocks ),
+			);
+		}
+	} else if ( has( transformation, 'apply' ) ) {
+		transformationResults = transformation.apply( [ firstBlock ] );
```

Comment:
To clarify: An `apply` implementation should always expect to receive an array of blocks (regardless of whether it's a `isMultiBlock` transform or not) and could return either an array of blocks or a single block object? Noting that this is slightly different than `transform`, since previously we assumed the function should both receive and return† blocks in array/single form as corresponding to whether `isMultiBlock` applies.

As implemented, it arguably makes `isMultiBlock` redundant, doesn't it? Not that I'm really opposed to that. Reflecting on all of the work we've done, a better implementation if we were to start fresh would probably just have been to always pass and always expect a return value of type array of block.

I'd suggest one of:

- Move the `has` condition to the first in the conditions, so we don't care whether `isMultiBlock` and just always pass `blocksArray`.
   - This would be tantamount to deprecating `isMultiBlock`
- Keep with the tradition the `isMultiBlock`, and change this line to pass `firstBlock` as the sole argument (as an object, not as an array)

Personally, I'm kinda leaning toward the second option, both to avoid another deprecation, and because singular transforms are common enough that a developer shouldn't need to always mess with `return blocks.map` for their singular block transforms.

```suggestion
		transformationResults = transformation.apply( firstBlock );
```

_† Technically the return type was never enforced one way or the other, since it's always cast to an array, but it might have been implied._

Reply:
> 'm kinda leaning toward the second option, both to avoid another deprecation, and because singular transforms are common enough that a developer shouldn't need to always mess with return blocks.map for their singular block transforms.

I've gone with the second approach.